src/pds/web_analytics/s3_sync.py

```diff
@@ -0,0 +1,427 @@
+                return True
+            if file_name == pattern:
+                return True
+            if file_name.endswith(pattern):
+                return True
+
+        return False
+
+    def run(self) -> None:
+        """Execute the sync process for all configured source paths."""
+        for src_path in self.src_paths.items():
+            self.sync_directory(src_path)
+
+    def sync_directory(self, path_tuple: Tuple[str, Dict[str, str]]) -> None:
+        """Sync a single directory to S3, including progress logging and deletion if specified.
+
+        Args:
+            path_tuple (tuple): A tuple containing the source path and include patterns.
+        """
+        src_path, path_include = path_tuple
+
+        # Ensure all files are gzipped before sync (if enabled)
+        if self.enable_gzip:
+            print(f"Ensuring files are gzipped in: {src_path}")
+            self.ensure_files_are_gzipped(src_path)
+        else:
+            print(f"Gzip compression disabled, syncing files as-is: {src_path}")
+
+        s3_base_path = os.path.join(self.s3_subdir, os.path.relpath(src_path, self.src_logdir))
+
+        # Collect all include patterns
+        all_patterns = []
+        for includes in path_include.values():
+            for pattern in includes:
+                # Update include patterns based on gzip setting
+                if self.enable_gzip and not pattern.endswith(".gz"):
+                    pattern += ".gz"
+                all_patterns.append(pattern)
+
+        # Upload files
+        uploaded_count = 0
+        total_files = 0
+
+        for root, _dirs, files in os.walk(src_path):
+            for file in files:
+                file_path = os.path.join(root, file)
+
+                if self.should_upload_file(file_path, all_patterns):
+                    total_files += 1
+
+                    # Calculate S3 key
+                    rel_path = os.path.relpath(file_path, src_path)
+                    s3_key = os.path.join(s3_base_path, rel_path).replace("\\", "/")
+
+                    print(f"Uploading: {file_path} -> s3://{self.bucket_name}/{s3_key}")
+
+                    if self.upload_file(file_path, s3_key):
+                        uploaded_count += 1
+
+                        # Delete source file if requested
+                        if self.delete:
+                            try:
+                                os.remove(file_path)
+                                print(f"Deleted source file: {file_path}")
+                            except Exception as e:
+                                print(f"Error deleting source file {file_path}: {str(e)}")
+
+        if uploaded_count > 0:
+            print(f"{src_path} sync to {s3_base_path}: {uploaded_count}/{total_files} files uploaded successfully.")
+        else:
+            print(f"{src_path} sync to {s3_base_path}: no files to upload.")
+
+    @staticmethod
+    def convert_size(size: int) -> str:
```

Comment:
The [humanize](https://pypi.org/project/humanize/)  API also has a nice function for this (`naturalsize`); but this works great too 👍